Thing.h

```diff
@@ -696,9 +692,11 @@ class ThingDevice {
   }
 
   void serialize(JsonObject descr, String ip, uint16_t port) {
-    descr["id"] = this->id;
+    descr["id"] = "uri:" + this->id;
     descr["title"] = this->title;
-    descr["@context"] = "https://webthings.io/schemas";
+
+    JsonArray context = descr.createNestedArray("@context");
+    context.add("https://www.w3.org/2019/wot/td/v1");
```

Comment:
I see that you removed the `https://webthings.io/schemas` `@context` from Thing Descriptions and replaced it with the W3C URL `https://www.w3.org./2019/wot/td/v1`. 

We are [planning to do this in the gateway](https://github.com/WebThingsIO/gateway/issues/2809) too once the gateway's Thing Descriptions are fully W3C compliant, but please note that in order for web things using the webthing-arduino library to use any of the capability schemas from `https://webthings.io/schemas/` then they should still include the webthings.io `@context`. This is an important part of the WebThings platform and if you remove these schemas then it will significantly degrade the user experience of users consuming these web things via WebThings Gateway, because it uses these semantic annotations to generate a richer UI. I suggest including both contexts in an array (e.g. `"@context": ["https://www.w3.org/2019/wot/td/v1", "https://webthings.io/schemas/"]`).

We could alternatively use a prefix (e.g. `"@context": ["https://www.w3.org/2019/wot/td/v1", {"wt": "https://webthings.io/schemas/"} ]`) but we'd need to add support for correctly consuming that in the gateway.

Reply:
It isn't JSON-LD compliant.
https://json-ld.org/playground/#startTab=tab-expanded&json-ld=%7B%22%40context%22%3A%5B%22https%3A%2F%2Fwebthings.io%2Fschemas%2F%22%5D%7D

Reply:
@benfrancis 
>That is certainly an option. Note that currently I think it's the case that WebThings Gateway would recognise the semantic annotations if both contexts were included anonymously, but unfortunately would not recognise them if a prefix is used. Therefore this option would require implementation work on WebThings Gateway in order for web things using this library to continue to benefit from the enhanced user experience these schemas provide.

I understand, but I think the effort would only result in 2-3 lines of new code, right? 

>I'm curious why JSON-LD parsers do not try to fetch prefix defined namespaces BTW?

Since it only need the corresponding namespaces, which is given by the prefix. When you start doing reasoning, then the processing engines need to fetch the content. 

>This is definitely a long term goal. However, the webthings.io schemas are currently used by around 100 gateway add-ons and an unknown number of web things built with the WebThings Framework.

The idea would be not to replace it immediately. I'm thinking about creating an RDF representation in parralel and doing a smooth integration in the future version. Nevertheless, my offer remains.

Reply:
> I understand, but I think the effort would only result in 2-3 lines of new code, right?

Hmm, if all we did was strip out the prefixes when consuming the Thing Description then yes, possibly. But if we want to support multiple schema repositories in the future (e.g. iot.schema.org) then, bearing in mind the gateway parses Thing Descriptions as JSON rather than JSON-LD, care would need to be taken to prevent name collisions all the way up the stack. The `@type` keywords from webthings.io/schemas are currently used in all the web components on the front end for example.

Unfortunately work on making the Consumer side of the gateway W3C compliant is currently paused, so this may have to wait, possibly until next year.

> The idea would be not to replace it immediately. I'm thinking about creating an RDF representation in parralel and doing a smooth integration in the future version. Nevertheless, my offer remains.

I do think creating an RDF representation of the schemas would be a worthwhile task, thank you. I could then separately figure out how to host it.

Reply:
>Hmm, if all we did was strip out the prefixes when consuming the Thing Description then yes, possibly.

I would also just see this as an interim solution until a RDF support of `https://webthings.io/schemas/` is available. Please note that I am trying to mitigate the situation here, as @Citrullin has made some efforts to make the code W3C-WoT compliant.  

> I do think creating an RDF representation of the schemas would be a worthwhile task, thank you. I could then separately figure out how to host it.

OK, I will check to see who can look at it.

Reply:
> Please note that I am trying to mitigate the situation here, as @Citrullin has made some efforts to make the code W3C-WoT compliant.

I see. Well if that is the only goal then as I mentioned, making the webthings.io/schemas URL resolve to a valid JSON-LD resource is not actually necessary to make the Thing Descriptions exposed by this library W3C compliant.

If your goal is to address the issues with this PR in order that it can land, that's really not the biggest problem in my view. The much bigger problem is the unnecessary and indiscriminate removal of all features which can't be described by a WoT Thing Description 1.0 Thing Description. That's inconsistent with the much more careful and considered approach we are taking elsewhere in the platform, which is something I've offered to discuss in more depth to ensure we are being consistent.

Reply:
> > Please note that I am trying to mitigate the situation here, as @Citrullin has made some efforts to make the code W3C-WoT compliant.
> 
> I see. Well if that is the only goal then as I mentioned, making the webthings.io/schemas URL resolve to a valid JSON-LD resource is not actually necessary to make the Thing Descriptions exposed by this library W3C compliant.

Prefix would be fine, since it doesn't result in a resolve. Please correct me if I am wrong, but that is not an option as well, because you need to implement something in the Gateway.

> If your goal is to address the issues with this PR in order that it can land, that's really not the biggest problem in my view. The much bigger problem is the unnecessary and indiscriminate removal of all features which can't be described by a WoT Thing Description 1.0 Thing Description. That's inconsistent with the much more careful and considered approach we are taking elsewhere in the platform, which is something I've offered to discuss in more depth to ensure we are being consistent.

Do you mean the endpoints /actions and /properties? If so, all the properties and actions are already listed in the TD itself. So, you would just duplicate the information. I mean fine with me to add it though. I don't have hard feelings about it.

Or do you to always return an array with multiple TDs in it? That wouldn't be TD compliant, even not with TD 1.1 as far as I know. And implementing TDD in Arduino would really be an overkill. 

Can you list your must haves in order to make it mergeable? I am getting really confused, because there are so many things you mentioned.

Reply:
> Prefix would be fine as well, since it doesn't result in a resolve. Please correct me if I am wrong, but that is not an option as well, because you need to implement something in the Gateway, right?

Yes, we just discussed that above.

> Do you mean the endpoints /actions and /properties? If so, all the properties and actions are already listed in the TD itself. So, you would just duplicate the information. I mean fine with me to add it though. I don't have hard feelings about it.
> 
> Or do you to always return an array with multiple TDs in it? That wouldn't be TD compliant, even not with TD 1.1 as far as I know. And implementing TDD in Arduino would really be an overkill.
> 
> Can you list your must haves in order to make it mergeable? I am getting really confused, because there are so many things you mentioned.

We've already discussed all of this, but I will summarise it all again below.

Reply:
I may have stumbled across a solution to the context issue in https://github.com/w3c/wot-thing-description/issues/442

Apparently because of the issues with content negotiation it is common practice to use one URI for the ontology which resolves to a human readable resource, then a second URI with a versioned subdirectory for the JSON-LD context.

On GitHub Pages it should be possible to use:
- `https://webthings.io/schemas` serving the current HTML resource
- `https://webthings.io/schemas/v1` serving a JSON resource

Although this means that implementations using the existing `https://webthings.io/schemas` context would still not resolve to a valid JSON-LD resource, new implementations could use the new context URI without breaking incoming links to the existing HTML page and I think the new URI could be used in WebThings Gateway without too much work.

This seems like a reasonable compromise to me, so if someone wants to provide a JSON-LD context file, then we can certainly host it at `https://webthings.io/schemas/v1`. I imagine that GitHub Pages will host it with an application/json content type, but hopefully that's not an issue.